File: binary_search_step.py
import sys

# Data class holding integer + string pair
class Data:
    def __init__(self, key, value):
        self.key = key
        self.value = value

    def __repr__(self):
        return f"{self.key}/{self.value}"

# Perform binary search on Data array and write the comparison steps
def binary_search(arr, target, writer):
    left = 0
    right = len(arr) - 1

    while left <= right:
        mid = left + (right - left) // 2

        # Write current comparison: row: key/value
        writer.write(f"{mid + 1}: {arr[mid].key}/{arr[mid].value}\n")

        if arr[mid].key == target:
            return mid  # Found, return index
        elif arr[mid].key < target:
            left = mid + 1
        else:
            right = mid - 1

    # Not found
    writer.write("-1\n")
    return -1

# Read the dataset (key/value pairs) from the provided CSV file
def read_data(filename):
    data = []
    with open(filename, 'r') as f:
        for line in f:
            parts = line.strip().split(",", 1)  # Split into key and value
            if len(parts) == 2:
                try:
                    key = int(parts[0].strip())
                    val = parts[1].strip()
                    data.append(Data(key, val))
                except ValueError:
                    print(f"Skipping malformed integer in line: {line}")
    return data

def main():
    if len(sys.argv) < 3:
        print("Usage: python3 binary_search_step.py <sorted_dataset_file> <target_integer>")
        return

    sorted_file = sys.argv[1]  # Expect the file name as an argument
    try:
        target = int(sys.argv[2])  # Target value for binary search
    except ValueError:
        print("Target must be an integer.")
        return

    try:
        # Read data from the CSV file (e.g., quick_sort_1000.csv)
        arr = read_data(sorted_file)
        if not arr:
            print("Dataset is empty.")
            return

        # Define output file name based on target value
        out_file = f"binary_search_step_{target}.txt"
        with open(out_file, 'w') as writer:
            found_index = binary_search(arr, target, writer)

        print(f"Binary search steps saved to {out_file}")

    except IOError as e:
        print(f"File error: {e}")

if __name__ == "__main__":
    main()

File: binary_search_step.java
import java.io.*;
import java.util.*;

public class binary_search_step {

    // Data class holding integer + string pair
    static class Data {
        int key;
        String value;
        Data(int k, String v) { key = k; value = v; }
    }

    // Perform binary search on Data array and write the comparison steps
    public static int binarySearch(Data[] arr, int target, BufferedWriter writer) throws IOException {
        int left = 0, right = arr.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            // Write current comparison: row: key/string
            writer.write(String.format("%d: %d/%s%n", mid + 1, arr[mid].key, arr[mid].value));

            if (arr[mid].key == target) {
                return mid; // Found, return index
            } else if (arr[mid].key < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        // Not found
        writer.write("-1\n");
        return -1;
    }

    // Read the dataset (key/value pairs) from the provided CSV file
    public static Data[] readData(String filename) throws IOException {
        List<Data> data = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",", 2);  // Split into key and value
                if (parts.length == 2) {
                    try {
                        int key = Integer.parseInt(parts[0].trim());
                        String val = parts[1].trim();
                        data.add(new Data(key, val));
                    } catch (NumberFormatException e) {
                        System.err.println("Skipping malformed integer in line: " + line);
                    }
                }
            }
        }
        return data.toArray(new Data[0]);
    }

    public static void main(String[] args) {
        if (args.length < 2) {
            System.out.println("Usage: java binary_search_step <sorted_dataset_file> <target_integer>");
            return;
        }

        String sortedFile = args[0]; // Expect the file name as an argument
        int target;

        try {
            target = Integer.parseInt(args[1]); // Target value for binary search
        } catch (NumberFormatException e) {
            System.err.println("Target must be an integer.");
            return;
        }

        try {
            // Read data from the CSV file (e.g., quick_sort_1000.csv)
            Data[] arr = readData(sortedFile);
            if (arr.length == 0) {
                System.err.println("Dataset is empty.");
                return;
            }

            // Define output file name based on target value
            String outFile = String.format("binary_search_step_%d.txt", target);
            try (BufferedWriter writer = new BufferedWriter(new FileWriter(outFile))) {
                int foundIndex = binarySearch(arr, target, writer);
                if (foundIndex >= 0) {
                    // Found target, index will be logged
                }
            }
            System.out.println("Binary search steps saved to " + outFile);
        } catch (IOException e) {
            System.err.println("File error: " + e.getMessage());
        }
    }
}
